src/derived/slot.rs

```diff
@@ -165,13 +166,13 @@ where
         // Check with an upgradable read to see if there is a value
         // already. (This permits other readers but prevents anyone
         // else from running `read_upgrade` at the same time.)
-        //
-        // FIXME(Amanieu/parking_lot#101) -- we are using a write-lock
-        // and not an upgradable read here because upgradable reads
-        // can sometimes encounter deadlocks.
-        let old_memo = match self.probe(db, self.state.write(), runtime, revision_now) {
+        let old_memo = match self.probe(db, self.state.upgradable_read(), runtime, revision_now) {
             ProbeState::UpToDate(v) => return v,
-            ProbeState::StaleOrAbsent(mut state) => {
+            ProbeState::StaleOrAbsent(state) => {
+                type RwLockUpgradableReadGuard<'a, T> =
+                    lock_api::RwLockUpgradableReadGuard<'a, RawRwLock, T>;
+
+                let mut state = RwLockUpgradableReadGuard::upgrade(state);
```

Comment:
This looks weird, but I couldn't figure out how to call this otherwise.

Reply:
Let's maybe just keep `.write` here, to not mix up library update with changes?

Reply:
Done.